### PR TITLE
Add unit tests for rule tags

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
None

## Summary
An inconsistency in tag formatting was found during a [rules PR to Kibana](https://github.com/elastic/kibana/pull/79061#discussion_r499857985). This adds two new unit tests for:

#### Casing and spacing
```
AssertionError: 594e0cbf-86cc-45aa-9ff7-ff27db27d3ed - AWS CloudTrail Log Created -> Invalid casing for expected tags
Actual tags: Sec Ops, ContinuousMonitoring
Expected tags: SecOps, Continuous Monitoring
```

#### Missing expected tags
(a rule with `index = ["winlogbeat-*", "logs-endpoint.events.*"]` and only `Elastic` tag)

```
AssertionError: 4630d948-40d4-4cef-ac69-4002e29bc3db - Adding Hidden File Attribute via Attrib -> Missing tags:
Actual tags: Elastic
Missing all of: Windows
Missing any of: MacOS, Linux
```

This was only based on known index <-> tag relationships based on a survey of the current rule set. We can expand this to include other relationships based on @seth-goodwin outcome on rule categorization (or save it for a separate PR).

As an example, if a Mitre mapping is present, ensure `Threat Detection` is a tag.

### CODE REVIEW

File: tests/test_all_rules.py [hunk 1]
```diff
@@ -186,3 +186,85 @@ def test_technique_deprecations(self):
...
+        expected_tags = [
+            'APM', 'AWS', 'Asset Visibility', 'Azure', 'Configuration Audit', 'Continuous Monitoring',
+            'Data Protection', 'Elastic', 'Endpoint', 'GCP', 'Identity and Access', 'Linux', 'Logging', 'ML', 'MacOS',
+            'Monitoring', 'Network', 'Okta', 'Packetbeat', 'Post-Execution', 'SecOps', 'Windows'
+        ]
+        expected_case = {strip_and_lower(t): t for t in expected_tags}
+
+        for rule in rules:
+            rule_tags = rule.contents.get('tags')
+            if rule_tags:
```
Comment (brokensound77) [inline tests/test_all_rules.py:210]: ready for python 3.8+

File: rules/network/discovery_post_exploitation_public_ip_reconnaissance.toml [hunk 2]
```diff
@@ -29,7 +29,7 @@ references = [
 risk_score = 21
 rule_id = "1d72d014-e2ab-4707-b056-9b96abe7b511"
 severity = "low"
-tags = ["Elastic", "Network", "Threat Detection, Preventing and Hunting", "Post-Execution"]
```
Comment (brokensound77) [inline rules/network/discovery_post_exploitation_public_ip_reconnaissance.toml:32]: These are inconsistent and should be fixed in a subsequent PR

File: rules/cross-platform/impact_hosts_file_modified.toml [hunk 3]
```diff
@@ -22,7 +22,7 @@ references = ["https://www.elastic.co/guide/en/beats/auditbeat/current/auditbeat
 risk_score = 47
 rule_id = "9c260313-c811-4ec8-ab89-8f6530e0246c"
 severity = "medium"
-tags = ["Elastic", "Linux", "Windows", "macOS"]
+tags = ["Elastic", "Linux", "Windows", "MacOS"]
```
Comment (rw-access) [inline rules/cross-platform/impact_hosts_file_modified.toml:25]: hold up. i think the original is way more correct.
just ask Apple

File: tests/test_all_rules.py [hunk 4]
```diff
@@ -186,3 +186,85 @@ def test_technique_deprecations(self):
...
+
+    def test_required_tags(self):
+        """Test that expected tags are present within rules."""
+        rules = rule_loader.load_rules().values()
+
+        # indexes considered; only those with obvious relationships included
+        # 'apm-*-transaction*', 'auditbeat-*', 'endgame-*', 'filebeat-*', 'logs-*', 'logs-aws*',
+        # 'logs-endpoint.alerts-*', 'logs-endpoint.events.*', 'logs-okta*', 'packetbeat-*', 'winlogbeat-*'
+
+        required_tags_map = {
```
Comment (rw-access) [inline tests/test_all_rules.py:228]: oh that's interesting

File: tests/test_all_rules.py [hunk 5]
```diff
@@ -186,3 +186,85 @@ def test_technique_deprecations(self):
...
+
+
+class TestRuleTags(unittest.TestCase):
+    """Test tags data for rules."""
+
+    def test_casing_and_spacing(self):
+        """Ensure consistent and expected casing for controlled tags."""
+        rules = rule_loader.load_rules().values()
+
+        def strip_and_lower(s):
```
Comment (rw-access) [inline tests/test_all_rules.py:198]: nit: `normalize`?

File: tests/test_all_rules.py [hunk 6]
```diff
@@ -186,3 +186,85 @@ def test_technique_deprecations(self):
...
+            return ''.join(s.lower().split())
+
+        expected_tags = [
+            'APM', 'AWS', 'Asset Visibility', 'Azure', 'Configuration Audit', 'Continuous Monitoring',
+            'Data Protection', 'Elastic', 'Endpoint', 'GCP', 'Identity and Access', 'Linux', 'Logging', 'ML', 'MacOS',
```
Comment (rw-access) [inline tests/test_all_rules.py:203]:
```suggestion
            'Data Protection', 'Elastic', 'Endpoint', 'GCP', 'Identity and Access', 'Linux', 'Logging', 'ML', 'macOS',
```

File: rules/cross-platform/impact_hosts_file_modified.toml (hunk 3 above)
Comment (brokensound77) [reply rules/cross-platform/impact_hosts_file_modified.toml:25]: wfm - so long as we are consistent

File: tests/test_all_rules.py (hunk 1 above)
Comment (brokensound77) [reply tests/test_all_rules.py:210]: `:=`

**Review by threat-punter — APPROVED** [review]

**Review by rw-access — APPROVED** [review]
works for me

**Pull request merged**